File: test/src/optimization/problem_spy_test.cpp
// Copyright (c) Sleipnir contributors

#include <stdint.h>

#include <format>
#include <fstream>
#include <string>

#include <catch2/catch_template_test_macros.hpp>
#include <catch2/catch_test_macros.hpp>
#include <sleipnir/autodiff/expression_type.hpp>
#include <sleipnir/optimization/problem.hpp>
#include <sleipnir/optimization/solver/exit_status.hpp>

#include "catch_matchers.hpp"
#include "catch_string_converters.hpp"
#include "coord.hpp"
#include "scalar_types_under_test.hpp"
#include "sleipnir/optimization/solver/iteration_info.hpp"

std::ifstream open(std::string_view filename) {
  std::ifstream file{std::string{filename}, std::ios::binary};
  if (!file.is_open()) {
    FAIL(std::format("Could not open {}", filename));
  }
  return file;
}

char read_char(std::ifstream& spy) {
  char value;
  spy.read(&value, 1);
  return value;
}

int32_t read_i32(std::ifstream& spy) {
  int32_t value;
  spy.read(reinterpret_cast<char*>(&value), 4);
  return value;
}

std::string read_str(std::ifstream& spy) {
  int32_t size = read_i32(spy);
  std::string value(size, '\0');
  spy.read(&value[0], size);
  return value;
}

Coord read_coord(std::ifstream& spy) {
  return Coord{read_i32(spy), read_i32(spy), read_char(spy)};
}

TEMPLATE_TEST_CASE("Problem - Spy", "[Problem]", SCALAR_TYPES_UNDER_TEST) {
  using T = TestType;

#ifdef SLEIPNIR_DISABLE_DIAGNOSTICS
  SKIP("Spy is disabled");
#endif

  slp::Problem<T> problem;

  auto x = problem.decision_variable();
  auto y = problem.decision_variable();
  x.set_value(T(20));
  y.set_value(T(20));

  problem.minimize(pow(x, T(4)) + pow(y, T(4)));

  problem.subject_to(x >= T(1));
  problem.subject_to(x <= T(10));
  problem.subject_to(y == T(2));

  int iterations = 0;
  problem.add_callback([&](const slp::IterationInfo<T>&) { ++iterations; });

  CHECK(problem.cost_function_type() == slp::ExpressionType::NONLINEAR);
  CHECK(problem.equality_constraint_type() == slp::ExpressionType::LINEAR);
  CHECK(problem.inequality_constraint_type() == slp::ExpressionType::LINEAR);

  CHECK(problem.solve({.diagnostics = true}, true) == slp::ExitStatus::SUCCESS);

  CHECK_THAT(x.value(), WithinAbs(T(1), T(1e-8)));
  CHECK_THAT(y.value(), WithinAbs(T(2), T(1e-8)));

  // Check H.spy
  {
    auto spy = open("H.spy");

    REQUIRE(read_str(spy) == "Hessian");             // Title
    REQUIRE(read_str(spy) == "Decision variables");  // Row label
    REQUIRE(read_str(spy) == "Decision variables");  // Col label
    REQUIRE(read_i32(spy) == 2);                     // Rows
    REQUIRE(read_i32(spy) == 2);                     // Cols

    // Coords
    for (int i = 0; i < iterations; ++i) {
      REQUIRE(read_i32(spy) == 2);  // Num coords
      CHECK(read_coord(spy) == Coord{0, 0, '+'});
      CHECK(read_coord(spy) == Coord{1, 1, '+'});
      if (spy.eof()) {
        FAIL("Reached end of file prematurely");
      }
    }
  }

  // Check A_e.spy
  {
    auto spy = open("A_e.spy");

    REQUIRE(read_str(spy) == "Equality constraint Jacobian");  // Title
    REQUIRE(read_str(spy) == "Constraints");                   // Row label
    REQUIRE(read_str(spy) == "Decision variables");            // Col label
    REQUIRE(read_i32(spy) == 1);                               // Rows
    REQUIRE(read_i32(spy) == 2);                               // Cols

    // Coords
    for (int i = 0; i < iterations; ++i) {
      REQUIRE(read_i32(spy) == 1);  // Num coords
      CHECK(read_coord(spy) == Coord{0, 1, '+'});
      if (spy.eof()) {
        FAIL("Reached end of file prematurely");
      }
    }
  }

  // Check A_i.spy
  {
    auto spy = open("A_i.spy");

    REQUIRE(read_str(spy) == "Inequality constraint Jacobian");  // Title
    REQUIRE(read_str(spy) == "Constraints");                     // Row label
    REQUIRE(read_str(spy) == "Decision variables");              // Col label
    REQUIRE(read_i32(spy) == 2);                                 // Rows
    REQUIRE(read_i32(spy) == 2);                                 // Cols

    // Coords
    for (int i = 0; i < iterations; ++i) {
      REQUIRE(read_i32(spy) == 2);  // Num coords
      CHECK(read_coord(spy) == Coord{0, 0, '+'});
      CHECK(read_coord(spy) == Coord{1, 0, '-'});
      if (spy.eof()) {
        FAIL("Reached end of file prematurely");
      }
    }
  }
}

File: python/test/optimization/problem_spy_test.py
import struct
from typing import BinaryIO

import pytest
from sleipnir.autodiff import ExpressionType
from sleipnir.optimization import ExitStatus, Problem


def read_char(spy: BinaryIO) -> bytes:
    return struct.unpack("c", spy.read(1))[0]


def read_i32(spy: BinaryIO) -> int:
    return struct.unpack("<i", spy.read(4))[0]


def read_str(spy: BinaryIO) -> str:
    size = read_i32(spy)
    return struct.unpack(f"{size}s", spy.read(size))[0].decode("utf-8")


class Coord:
    def __init__(self, row: int, col: int, sign: bytes):
        self.row = row
        self.col = col
        self.sign = sign

    def __eq__(self, other) -> bool:
        return (
            self.row == other.row and self.col == other.col and self.sign == other.sign
        )


def read_coord(spy: BinaryIO) -> Coord:
    return Coord(read_i32(spy), read_i32(spy), read_char(spy))


iterations = 0


def test_problem_spy():
    problem = Problem()

    x, y = problem.decision_variable(2)
    x.set_value(20.0)
    y.set_value(20.0)

    problem.minimize(x**4 + y**4)

    problem.subject_to(x >= 1.0)
    problem.subject_to(x <= 10.0)
    problem.subject_to(y == 2.0)

    global iterations
    iterations = 0

    def callback(info):
        global iterations
        iterations += 1
        return False

    problem.add_callback(callback)

    assert problem.cost_function_type() == ExpressionType.NONLINEAR
    assert problem.equality_constraint_type() == ExpressionType.LINEAR
    assert problem.inequality_constraint_type() == ExpressionType.LINEAR

    assert problem.solve(diagnostics=True, spy=True) == ExitStatus.SUCCESS

    assert x.value() == pytest.approx(1.0, abs=1e-8)
    assert y.value() == pytest.approx(2.0, abs=1e-8)

    # Check H.spy
    with open("H.spy", mode="rb") as spy:
        assert read_str(spy) == "Hessian"  # Title
        assert read_str(spy) == "Decision variables"  # Row label
        assert read_str(spy) == "Decision variables"  # Col label
        assert read_i32(spy) == 2  # Rows
        assert read_i32(spy) == 2  # Cols

        # Coords
        for _ in range(iterations):
            assert read_i32(spy) == 2  # Num coords
            assert read_coord(spy) == Coord(0, 0, b"+")
            assert read_coord(spy) == Coord(1, 1, b"+")

    # Check A_e.spy
    with open("A_e.spy", mode="rb") as spy:
        assert read_str(spy) == "Equality constraint Jacobian"  # Title
        assert read_str(spy) == "Constraints"  # Row label
        assert read_str(spy) == "Decision variables"  # Col label
        assert read_i32(spy) == 1  # Rows
        assert read_i32(spy) == 2  # Cols

        # Coords
        for _ in range(iterations):
            assert read_i32(spy) == 1  # Num coords
            assert read_coord(spy) == Coord(0, 1, b"+")

    # Check A_i.spy
    with open("A_i.spy", mode="rb") as spy:
        assert read_str(spy) == "Inequality constraint Jacobian"  # Title
        assert read_str(spy) == "Constraints"  # Row label
        assert read_str(spy) == "Decision variables"  # Col label
        assert read_i32(spy) == 2  # Rows
        assert read_i32(spy) == 2  # Cols

        # Coords
        for _ in range(iterations):
            assert read_i32(spy) == 2  # Num coords
            assert read_coord(spy) == Coord(0, 0, b"+")
            assert read_coord(spy) == Coord(1, 0, b"-")
